lib/Performance/Probability.pm

```diff
@@ -129,54 +129,139 @@ sub _variance_x_square {
-                my $max_start_time = $start_time->[$i] > $start_time->[$j] ? $start_time->[$i] : $start_time->[$j];
-                my $b_interval     = $min_end_time - $max_start_time;
+            for ($j = 0; $j < @{$sell_time}; ++$j) {
+                if ($i != $j and $underlying->[$i] eq $underlying->[$j]) {
 
-                if ($b_interval > 0) {
+                    #check for time overlap.
+                    my $min_end_time   = $sell_time->[$i] < $sell_time->[$j]   ? $sell_time->[$i]  : $sell_time->[$j];
+                    my $max_start_time = $start_time->[$i] > $start_time->[$j] ? $start_time->[$i] : $start_time->[$j];
+                    my $b_interval     = $min_end_time - $max_start_time;
 
-                    #calculate first and second contracts durations. please see the documentation for details
+                    if ($b_interval > 0) {
 
-                    my $first_contract_duration  = ($sell_time->[$i] - $start_time->[$i]);
-                    my $second_contract_duration = ($sell_time->[$j] - $start_time->[$j]);
+                        #calculate first and second contracts durations. please see the documentation for details
 
-                    my $i_strike = 0.0 - Math::Gauss::XS::inv_cdf($pk->[$i]);
-                    my $j_strike = 0.0 - Math::Gauss::XS::inv_cdf($pk->[$j]);
+                        my $first_contract_duration  = ($sell_time->[$i] - $start_time->[$i]);
+                        my $second_contract_duration = ($sell_time->[$j] - $start_time->[$j]);
 
-                    my $corr_ij = $b_interval / (sqrt($first_contract_duration) * sqrt($second_contract_duration));
+                        my $i_strike = 0.0 - Math::Gauss::XS::inv_cdf($pk->[$i]);
+                        my $j_strike = 0.0 - Math::Gauss::XS::inv_cdf($pk->[$j]);
 
-                    if ($types->[$i] ne $types->[$j]) {
-                        $corr_ij = -1 * $corr_ij;
-                    }
+                        my $corr_ij = $b_interval / (sqrt($first_contract_duration) * sqrt($second_contract_duration));
 
-                    if ($corr_ij < -1 or $corr_ij > 1) {
-                        next;
-                    }
+                        if ($types->[$i] ne $types->[$j]) {
+                            $corr_ij = -1 * $corr_ij;
+                        }
+
+                        if ($corr_ij < -1 or $corr_ij > 1) {
+                            next;
+                        }
+
+                        my $p_ij = Math::BivariateCDF::bivnor($i_strike, $j_strike, $corr_ij);
 
-                    my $p_ij = Math::BivariateCDF::bivnor($i_strike, $j_strike, $corr_ij);
+                        my $covariance_ij =
+                            ($p_ij - $pk->[$i] * $pk->[$j]) * ($wk->[$i] - $lk->[$i]) * ($wk->[$j] - $lk->[$j]);
+                        $covariance = $covariance + $covariance_ij;
+                    }
+                }
+            }
+        } else {
```

Comment:
I think it might be good to add an else if condition here to check if it is a digit contract